debezium-server/debezium-server-rocketmq/src/main/java/io/debezium/server/rocketmq/RocketMqChangeConsumer.java

```diff
@@ -0,0 +1,160 @@
+                            config.getValue(PROP_PRODUCER_ACCESS_KEY, String.class),
+                            config.getValue(PROP_PRODUCER_SECRET_KEY, String.class)));
+        }
+        this.mqProducer = new DefaultMQProducer(rpcHook);
+        this.mqProducer.setNamesrvAddr(config.getValue(PROP_PRODUCER_NAME_SRV_ADDR, String.class));
+        this.mqProducer.setInstanceName(createUniqInstance(config.getValue(PROP_PRODUCER_NAME_SRV_ADDR, String.class)));
+        this.mqProducer.setProducerGroup(config.getValue(PROP_PRODUCER_GROUP, String.class));
+
+        if (config.getOptionalValue(PROP_PRODUCER_SEND_MSG_TIMEOUT, Integer.class).isPresent()) {
+            this.mqProducer.setSendMsgTimeout(config.getValue(PROP_PRODUCER_SEND_MSG_TIMEOUT, Integer.class));
+        }
+
+        if (config.getOptionalValue(PROP_PRODUCER_MAX_MESSAGE_SIZE, Integer.class).isPresent()) {
+            this.mqProducer.setMaxMessageSize(config.getValue(PROP_PRODUCER_MAX_MESSAGE_SIZE, Integer.class));
+        }
+
+        this.mqProducer.setLanguage(LanguageCode.JAVA);
+        startProducer();
+    }
+
+    private void startProducer() {
+        try {
+            this.mqProducer.start();
+            LOGGER.info("Consumer started...");
+        } catch (MQClientException e) {
+            throw new DebeziumException(e);
+        }
+    }
+
+    private String createUniqInstance(String prefix) {
+        return prefix.concat("-").concat(UUID.randomUUID().toString());
+    }
+
+    @PreDestroy
+    void close() {
+        // Closed rocketmq producer
+        LOGGER.info("Consumer destroy...");
+        if (mqProducer != null) {
+            mqProducer.shutdown();
+        }
+    }
+
+    @Override
+    public void handleBatch(List<ChangeEvent<Object, Object>> records, DebeziumEngine.RecordCommitter<ChangeEvent<Object, Object>> committer)
+            throws InterruptedException {
+        final CountDownLatch latch = new CountDownLatch(records.size());
+        for (ChangeEvent<Object, Object> record : records) {
+            try {
+                final String topicName = streamNameMapper.map(record.destination());
+                String key = getString(record.key());
+                mqProducer.send(new Message(topicName, null, key, getBytes(record.value())), new SelectMessageQueueByHash(), key, new SendCallback() {
+                    @Override
+                    public void onSuccess(SendResult sendResult) {
+                        LOGGER.debug("Sent message with offset: {}", sendResult.getQueueOffset());
+                        latch.countDown();
+                    }
+
+                    @Override
+                    public void onException(Throwable throwable) {
+                        LOGGER.error("Failed to send record to {}:", record.destination(), throwable);
+                        throw new DebeziumException(throwable);
+                    }
+                });
+                committer.markProcessed(record);
```

Comment:
>> I already mentioned, that moving markProcessed() to onSuccess() was wrong proposal and you are right that >> it will mess up the order. But I still believe that you should markProcessed() only when you are sure that the >> messages was delivered, i.e. after latch.await()

> OK, I have rollbacked to the original version

as mentioned above, I still believe that you should call `markProcessed()` only when you are sure that the messages was delivered, i.e. after `latch.await()`. If you want to do it this way, are you sure it cannot cause any issue while waiting for `latch.await()` (i.e. during this time Kafka will think that these events are processed)?

Reply:
> it cannot cause any issue

You are right. Although the logic is simple, it may still cause some problems

Reply:
> > > I already mentioned, that moving markProcessed() to onSuccess() was wrong proposal and you are right that >> it will mess up the order. But I still believe that you should markProcessed() only when you are sure that the >> messages was delivered, i.e. after latch.await()
> 
> > OK, I have rollbacked to the original version
> 
> as mentioned above, I still believe that you should call `markProcessed()` only when you are sure that the messages was delivered, i.e. after `latch.await()`. If you want to do it this way, are you sure it cannot cause any issue while waiting for `latch.await()` (i.e. during this time Kafka will think that these events are processed)?

You are right. Although the logic is simple, it may still cause some problems